[PlayStation][wincairo] fix assertion with SingleWebProcess mode

haruhisa-shin opened the pull request:
#### 5d53332466c251054cb0579d7916b764ee11db0b
<pre>
[PlayStation][wincairo] fix assertion with SingleWebProcess mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=263185">https://bugs.webkit.org/show_bug.cgi?id=263185</a>

Reviewed by Chris Dumez.

The WebPageProxy::continueNavigationInNewProcess() function attempts to
create a new WebProcess to continue processing a response from a
particular websites.
If the application sets usesSingleWebProcess = true, the
WebPageProxy will attempt to reuse an existing process, but
unfortunately there is an assert that is not supposed to do that,
so the application with debug build will crash.

The NetworkResourceLoader in NetworkProcess decides whether to create
a new WebProcess.
So this patch makes two changes:
1. add SingleWebProcess flag to NetworkProcessPreferencesForWebProcess
2. use StayInGroup(do not create a new WebProcess) if SingleWebProcess

Incidentally, besides some ASSERTs, the connection with the automation
target was also not very well considered for SingleWebProcess mode.
For example, WebPageProxy::resetStateAfterProcessTermination() calls
disconnect() to target, but no one reconnects afterwards.
This raises a WebDriver Exception even if it is a release build.
This patch will solve this as well.

Canonical link: <a href="https://commits.webkit.org/278467@main">https://commits.webkit.org/278467@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2309d3c8270f3df82572ed0ea5d75dba018903f6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49677 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28965 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52726 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52919 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/353 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34985 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26582 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40532 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51777 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26494 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42790 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21652 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23936 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8048 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45852 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44470 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54504 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24764 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20919 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47914 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26031 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42897 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46947 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11096 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26879 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25756 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4244,7 +4244,7 @@ void WebPageProxy::commitProvisionalPage(FrameIdentifier frameID, FrameInfoData&
     RefPtr<WebFrameProxy> mainFrameInPreviousProcess = m_mainFrame;
 
-    ASSERT(m_process.ptr() != &m_provisionalPage->process());
+    ASSERT(m_process->processPool().configuration().usesSingleWebProcess() || m_process.ptr() != &m_provisionalPage->process());
```
I don't think we should have ever ended up this code path in single process mode. This code path occurs after process-swapping on navigation. When you're using a single WebProcess, you cannot process swap.

Odds are that the proper fix should happen earlier, to make sure we don't end up creating a provisional page in the first place.

haruhisa-shin replied:
@cdumez Thanks for the review. Also, I am sorry for the delay in responding.

> Odds are that the proper fix should happen earlier, to make sure we don't end up creating a provisional page in the first place.

To avoid creating a provisional page, it may be necessary to modify the NetworkProcess side.
The value of the `browserContextGroupSwitchDecision` determined at [NetworkResourceLoader::sendDidReceiveResponsePotentiallyInNewBrowsingContextGroup()](https://github.com/WebKit/WebKit/blob/main/Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp#L993) decides whether process swapping was executed or not.

I will try to create a modification in that the browserContextGroupSwitchDecision is `StayInGroup` in the case of SingleWebProcess to avoid creating a provisional page.

rkirsling commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -1000,7 +1000,7 @@ void NetworkResourceLoader::didReceiveResponse(ResourceResponse&& receivedRespon
 void NetworkResourceLoader::sendDidReceiveResponsePotentiallyInNewBrowsingContextGroup(const WebCore::ResourceResponse& response, PrivateRelayed privateRelayed, bool needsContinueDidReceiveResponseMessage)
 {
-    auto browsingContextGroupSwitchDecision = toBrowsingContextGroupSwitchDecision(m_currentCoopEnforcementResult);
+    auto browsingContextGroupSwitchDecision = m_connection->networkProcess().usesSingleWebProcess()? BrowsingContextGroupSwitchDecision::StayInGroup: toBrowsingContextGroupSwitchDecision(m_currentCoopEnforcementResult);
```
I'm surprised the style checker didn't catch this one, but:
```suggestion
    auto browsingContextGroupSwitchDecision = m_connection->networkProcess().usesSingleWebProcess() ? BrowsingContextGroupSwitchDecision::StayInGroup : toBrowsingContextGroupSwitchDecision(m_currentCoopEnforcementResult);
```

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.serialization.in:
```diff
@@ -56,6 +56,7 @@ headers: "NetworkProcessCreationParameters.h" "AuxiliaryProcessCreationParameter
 #if ENABLE(BUILT_IN_NOTIFICATIONS)
     bool builtInNotificationsEnabled
 #endif
+    bool usesSingleWebProcess
```
For Cocoa ports, the network process is a singleton. However, the `usesSingleWebProcess` is per ProcessPool in the UIProcess. As a result, having a global `usesSingleWebProcess` for the whole network process isn't quite right.

Once way to address this would be to move `usesSingleWebProcess` to `NetworkProcessPreferencesForWebProcess` instead. I'm trying to think if there is an easier way but at least having the flag on `NetworkProcessPreferencesForWebProcess` would be correct.

cdumez reviewed (changes requested):
I think we should move this flag to NetworkProcessPreferencesForWebProcess instead so that it can be per WebProcess instead of NetworkProcess global.

haruhisa-shin committed:
[PlayStation][wincairo] fix assertion with SingleWebProcess mode
https://bugs.webkit.org/show_bug.cgi?id=263185

Reviewed by Chris Dumez.

The WebPageProxy::continueNavigationInNewProcess() function attempts to
create a new WebProcess to continue processing a response from a
particular websites.
If the application sets usesSingleWebProcess = true, the
WebPageProxy will attempt to reuse an existing process, but
unfortunately there is an assert that is not supposed to do that,
so the application with debug build will crash.

The NetworkResourceLoader in NetworkProcess decides whether to create
a new WebProcess.
So this patch makes two changes:
1. add SingleWebProcess flag to NetworkProcessPreferencesForWebProcess
2. use StayInGroup(do not create a new WebProcess) if SingleWebProcess

Incidentally, besides some ASSERTs, the connection with the automation
target was also not very well considered for SingleWebProcess mode.
For example, WebPageProxy::resetStateAfterProcessTermination() calls
disconnect() to target, but no one reconnects afterwards.
This raises a WebDriver Exception even if it is a release build.
This patch will solve this as well.

Canonical link: https://commits.webkit.org/278467@main
